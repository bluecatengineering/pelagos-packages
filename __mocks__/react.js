module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
