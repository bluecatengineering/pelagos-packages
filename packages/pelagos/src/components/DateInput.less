@import '~@bluecateng/pelagos/less/elevations';
@import '~@bluecateng/pelagos/less/focus';
@import '~@bluecateng/pelagos/less/spacing';

@layer pelagos {
	.DateInput {
		position: relative;
		flex-direction: row;
		overflow: hidden;

		&__input {
			width: 0;
			flex: 1;
			padding-right: @sp-48;

			&[aria-invalid='true'] {
				outline: 2px solid var(--support-error);
				outline-offset: -2px;

				&:focus {
					@focus-ring-error();
				}
			}
		}

		&__icon {
			position: absolute;
			top: 0;
			right: 0;
		}

		&__calendar {
			@elv-04();
			position: absolute;
			z-index: 1000;
		}
	}
}
