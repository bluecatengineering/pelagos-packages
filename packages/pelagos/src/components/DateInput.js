import {useCallback, useEffect, useRef, useState} from 'react';
import {createPortal} from 'react-dom';
import PropTypes from 'prop-types';
import {createFocusTrap} from 'focus-trap';
import {t} from '@bluecateng/l10n.macro';
import CalendarIcon from '@carbon/icons-react/es/Calendar';

import useLayer from '../hooks/useLayer';

import Layer from './Layer';
import Calendar from './Calendar';
import IconButton from './IconButton';

import './DateInput.less';

const DateInput = ({className, value, disabled, error, format, parse, onChange, ...props}) => {
	const wrapperRef = useRef(null);
	const popUpRef = useRef(null);

	const level = useLayer();

	const [calendarTime, setCalendarTime] = useState(null);
	const handleInputChange = useCallback((event) => onChange(event.target.value), [onChange]);
	const handleButtonClick = useCallback(() => {
		// this is called only when the calendar is *not* visible
		setCalendarTime(parse(value) ?? new Date().setHours(0, 0, 0, 0));
	}, [parse, value]);
	const handleCalendarChange = useCallback(
		(time) => {
			onChange(format(time));
			setCalendarTime(null);
		},
		[format, onChange]
	);

	useEffect(() => {
		if (calendarTime === null) {
			return undefined;
		}

		const wrapper = wrapperRef.current;
		const {bottom, left} = wrapper.getBoundingClientRect();

		const popUp = popUpRef.current;
		popUp.style.top = `${bottom}px`;
		popUp.style.left = `${left}px`;

		const trap = createFocusTrap(popUp, {
			initialFocus: '.Calendar [tabindex="0"]',
			allowOutsideClick: (event) => {
				if (event.type === 'click') {
					trap.deactivate();
				}
				return false;
			},
			onDeactivate: () => setCalendarTime(null),
		});
		trap.activate();
		return trap.deactivate;
	}, [calendarTime]);

	return (
		<div className={`DateInput${className ? ` ${className}` : ''}`} ref={wrapperRef}>
			<Layer
				{...props}
				as="input"
				className="DateInput__input"
				type="text"
				value={value}
				disabled={disabled}
				aria-invalid={error}
				onChange={handleInputChange}
			/>
			<IconButton
				className="DateInput__icon"
				icon={CalendarIcon}
				disabled={disabled}
				aria-label={value ? t`Change date, ${value}` : t`Choose date`}
				onClick={handleButtonClick}
			/>
			{calendarTime !== null &&
				createPortal(
					<Layer className="DateInput__popUp" level={level + 1} ref={popUpRef}>
						<Calendar value={calendarTime} onChange={handleCalendarChange} />
					</Layer>,
					document.body
				)}
		</div>
	);
};

DateInput.propTypes = {
	/** The component class name(s). */
	className: PropTypes.string,
	/** The field value. */
	value: PropTypes.string,
	/** The placeholder text. */
	placeholder: PropTypes.string,
	/** Whether the field should be focused on display. */
	autoFocus: PropTypes.bool,
	/** Whether the field is disabled. */
	disabled: PropTypes.bool,
	/** Whether the component is in error. */
	error: PropTypes.bool,
	/** Function invoked to format the date selected form the calendar. */
	format: PropTypes.func,
	/** Function invoked to parse the value to pass to the calendar, must return null if the value is not valid. */
	parse: PropTypes.func,
	/** Function invoked when the value changes. */
	onChange: PropTypes.func,
};

export default DateInput;
