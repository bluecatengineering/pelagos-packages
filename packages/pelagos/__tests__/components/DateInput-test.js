import {useEffect, useRef, useState} from 'react';
import {shallow} from 'enzyme';
import {createFocusTrap} from 'focus-trap';

import DateInput from '../../src/components/DateInput';

jest.unmock('../../src/components/DateInput');

const anyFunction = expect.any(Function);

const body = {tag: 'body'};
global.document = {body};

describe('DateInput', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<DateInput />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(<DateInput className="TestClass" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when value is set', () => {
			const wrapper = shallow(<DateInput value="foo" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when calendarTime is set', () => {
			useState.mockReturnValueOnce([1000]);
			const wrapper = shallow(<DateInput />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onChange when the input changes', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<DateInput onChange={onChange} />);
			wrapper.find('[as="input"]').simulate('change', {target: {value: 'foo'}});
			expect(onChange.mock.calls).toEqual([['foo']]);
		});

		it('calls setCalendarTime when the button is clicked', () => {
			const setCalendarTime = jest.fn();
			const parse = jest.fn().mockReturnValue(1000);
			useState.mockReturnValueOnce([null, setCalendarTime]);
			const wrapper = shallow(<DateInput value="foo" parse={parse} />);
			wrapper.find('IconButton').simulate('click');
			expect(parse.mock.calls).toEqual([['foo']]);
			expect(setCalendarTime.mock.calls).toEqual([[1000]]);
		});

		it('calls setCalendarTime when the button is clicked and parse returns null', () => {
			const setCalendarTime = jest.fn();
			const parse = jest.fn();
			const setHours = jest.fn().mockReturnValue(2000);
			useState.mockReturnValueOnce([null, setCalendarTime]);
			jest.spyOn(global, 'Date').mockImplementation(() => ({setHours}));
			const wrapper = shallow(<DateInput value="foo" parse={parse} />);
			wrapper.find('IconButton').simulate('click');
			expect(parse.mock.calls).toEqual([['foo']]);
			expect(Date.mock.calls).toEqual([[]]);
			expect(setHours.mock.calls).toEqual([[0, 0, 0, 0]]);
			expect(setCalendarTime.mock.calls).toEqual([[2000]]);
		});

		it('calls onChange when the Calendar calls onChange', () => {
			const setCalendarTime = jest.fn();
			const format = jest.fn().mockReturnValue('bar');
			const onChange = jest.fn();
			useState.mockReturnValueOnce([1000, setCalendarTime]);
			const wrapper = shallow(<DateInput format={format} onChange={onChange} />);
			wrapper.find('Calendar').simulate('change', 'foo');
			expect(format.mock.calls).toEqual([['foo']]);
			expect(onChange.mock.calls).toEqual([['bar']]);
			expect(setCalendarTime.mock.calls).toEqual([[null]]);
		});

		it('adds an effect which creates and activates a focus trap', () => {
			const wrapper = {
				closest: jest.fn().mockReturnValue({dataset: {layer: '2'}}),
				getBoundingClientRect: jest.fn().mockReturnValue({bottom: 100, left: 200}),
			};
			const popUp = {style: {}};
			const setCalendarTime = jest.fn();
			const activate = jest.fn();
			const deactivate = jest.fn();
			useRef.mockReturnValueOnce({current: wrapper}).mockReturnValueOnce({current: popUp});
			useState.mockReturnValueOnce([1000, setCalendarTime]);
			createFocusTrap.mockReturnValue({activate, deactivate});
			shallow(<DateInput className="TestClass" />);
			expect(useEffect.mock.calls[0]).toEqual([anyFunction, [1000]]);
			expect(useEffect.mock.calls[0][0]()).toBe(deactivate);
			expect(activate.mock.calls).toEqual([[]]);
			expect(createFocusTrap.mock.calls).toEqual([
				[popUp, {initialFocus: '.Calendar [tabindex="0"]', allowOutsideClick: anyFunction, onDeactivate: anyFunction}],
			]);

			const {allowOutsideClick, onDeactivate} = createFocusTrap.mock.calls[0][1];

			expect(allowOutsideClick({type: 'foo'})).toBe(false);
			expect(deactivate.mock.calls).toEqual([]);

			expect(allowOutsideClick({type: 'click'})).toBe(false);
			expect(deactivate.mock.calls).toEqual([[]]);

			onDeactivate();
			expect(setCalendarTime.mock.calls).toEqual([[null]]);
		});

		it('adds an effect which does not create a focus trap when calendarTime is null', () => {
			const button = {};
			const popUp = {};
			useRef.mockReturnValueOnce({current: button}).mockReturnValueOnce({current: popUp});
			useState.mockReturnValueOnce([null]);
			shallow(<DateInput className="TestClass" />);
			expect(useEffect.mock.calls[0]).toEqual([anyFunction, [null]]);
			expect(useEffect.mock.calls[0][0]()).toBeUndefined();
			expect(createFocusTrap.mock.calls).toEqual([]);
		});
	});
});
