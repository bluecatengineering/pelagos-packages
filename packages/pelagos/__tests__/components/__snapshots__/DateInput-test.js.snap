// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput rendering renders expected elements 1`] = `
<Layer
  className="DateInput"
>
  <input
    className="DateInput__input"
    onChange={[Function]}
    type="text"
  />
  <IconButton
    aria-label="Choose date"
    className="DateInput__icon"
    icon={
      {
        "icon": [],
        "iconName": "calendar",
        "prefix": "far",
      }
    }
    onClick={[Function]}
    size="medium"
    tooltipPlacement="right"
    type="ghost"
  />
</Layer>
`;

exports[`DateInput rendering renders expected elements when calendarTime is set 1`] = `
<Layer
  className="DateInput"
>
  <input
    className="DateInput__input"
    onChange={[Function]}
    type="text"
  />
  <IconButton
    aria-label="Choose date"
    className="DateInput__icon"
    icon={
      {
        "icon": [],
        "iconName": "calendar",
        "prefix": "far",
      }
    }
    onClick={[Function]}
    size="medium"
    tooltipPlacement="right"
    type="ghost"
  />
  <React.Portal
    container={
      {
        "tag": "body",
      }
    }
  >
    <Calendar
      className="DateInput__calendar"
      onChange={[Function]}
      value={1000}
    />
  </React.Portal>
</Layer>
`;

exports[`DateInput rendering renders expected elements when className is set 1`] = `
<Layer
  className="DateInput TestClass"
>
  <input
    className="DateInput__input"
    onChange={[Function]}
    type="text"
  />
  <IconButton
    aria-label="Choose date"
    className="DateInput__icon"
    icon={
      {
        "icon": [],
        "iconName": "calendar",
        "prefix": "far",
      }
    }
    onClick={[Function]}
    size="medium"
    tooltipPlacement="right"
    type="ghost"
  />
</Layer>
`;

exports[`DateInput rendering renders expected elements when value is set 1`] = `
<Layer
  className="DateInput"
>
  <input
    className="DateInput__input"
    onChange={[Function]}
    type="text"
    value="foo"
  />
  <IconButton
    aria-label="Change date, foo"
    className="DateInput__icon"
    icon={
      {
        "icon": [],
        "iconName": "calendar",
        "prefix": "far",
      }
    }
    onClick={[Function]}
    size="medium"
    tooltipPlacement="right"
    type="ghost"
  />
</Layer>
`;
